zeebe/dynamic-config/src/test/java/io/camunda/zeebe/dynamic/config/api/ClusterPatchRequestTransformerTest.java

```diff
@@ -0,0 +1,412 @@
+      @ForAll @IntRange(min = 1, max = 5) final int oldReplicationFactor,
+      @ForAll @IntRange(min = 1, max = 5) final int newReplicationFactor) {
+    shouldScaleBrokersAndPartitionsByCount(
+        oldPartitionCount,
+        Optional.of(newPartitionCount),
+        oldReplicationFactor,
+        Optional.of(newReplicationFactor),
+        oldClusterSize,
+        Optional.of(newClusterSize));
+  }
+
+  void shouldScaleBrokersAndPartitionsByCount(
+      final int oldPartitionCount,
+      final Optional<Integer> newPartitionCount,
+      final int oldReplicationFactor,
+      final Optional<Integer> newReplicationFactor,
+      final int oldClusterSize,
+      final Optional<Integer> newClusterSize) {
+    // given
+    final var expectedNewDistribution =
+        new RoundRobinPartitionDistributor()
+            .distributePartitions(
+                getClusterMembers(newClusterSize.orElse(oldClusterSize)),
+                getSortedPartitionIds(newPartitionCount.orElse(oldPartitionCount)),
+                newReplicationFactor.orElse(oldReplicationFactor));
+
+    final var oldDistribution =
+        new RoundRobinPartitionDistributor()
+            .distributePartitions(
+                getClusterMembers(oldClusterSize),
+                getSortedPartitionIds(oldPartitionCount),
+                oldReplicationFactor);
+    ClusterConfiguration oldClusterTopology =
+        ConfigurationUtil.getClusterConfigFrom(true, oldDistribution, partitionConfig);
+    for (final MemberId member : getClusterMembers(oldClusterSize)) {
+      if (!oldClusterTopology.hasMember(member)) {
+        oldClusterTopology =
+            oldClusterTopology.addMember(member, MemberState.initializeAsActive(Map.of()));
+      }
+    }
+
+    // when
+    final var patchRequest =
+        new ClusterPatchRequest(
+            newClusterSize,
+            Set.of(),
+            Set.of(),
+            newPartitionCount,
+            newReplicationFactor,
+            false,
+            false);
+
+    applyRequestAndVerifyResultingTopology(
+        newPartitionCount.orElse(oldPartitionCount),
+        getClusterMembers(newClusterSize.orElse(oldClusterSize)),
+        patchRequest,
+        oldClusterTopology,
+        expectedNewDistribution);
+  }
+
+  private void applyRequestAndVerifyResultingTopology(
+      final int partitionCount,
+      final Set<MemberId> expectedMembers,
+      final ClusterPatchRequest patchRequest,
+      final ClusterConfiguration oldClusterTopology,
+      final Set<PartitionMetadata> expectedNewDistribution) {
+
+    // when
+    final var result =
+        new ClusterPatchRequestTransformer(patchRequest, MemberId.from("0"))
+            .operations(oldClusterTopology);
+    assertThat(result).isRight();
+    final var operations = result.get();
+
+    // apply operations to generate new topology
+    final ClusterConfiguration newTopology =
+        TestTopologyChangeSimulator.apply(oldClusterTopology, operations);
```

Comment:
:thought_balloon: just in theory, we could do the same in the request transformer to validate the request, right? That way we'd ensure that even in production, the generated operations always lead to the expected outcome.

Reply:
In the tests we can verify it against an expected outcome. But in production, we don't know the expected outcome. 

We are validating generated operation in the `ConfigurationChangeCoordinator` by a simulator. May be that's what you meant.

Reply:
Oh yeah of course, we only have the patch request and not the full information to calculate the expected topology at the end. Nevermind then :+1: